cpp/src/reshape/table_to_array.cu

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/reshape.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cub/device/device_memcpy.cuh>
+#include <cuda_runtime.h>
+#include <thrust/device_vector.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+namespace cudf {
+namespace {
+
+template <typename T>
+void _table_to_device_array(cudf::table_view const& input,
+                            void* output,
+                            rmm::cuda_stream_view stream)
+{
+  auto const num_columns = input.num_columns();
+  auto const num_rows    = input.num_rows();
+  auto const item_size   = sizeof(T);
+
+  std::vector<void*> dsts(num_columns);
+  std::vector<void const*> srcs(num_columns);
+  std::vector<size_t> sizes(num_columns, item_size * num_rows);
+
+  auto* base_ptr = static_cast<uint8_t*>(output);
```

Comment:
For "raw bytes" pointers like this I think `cuda::std::byte` could be a better option.
```suggestion
  auto* base_ptr = static_cast<cuda::std::byte*>(output);
```